src/java.base/share/classes/jdk/internal/ref/PhantomCleanable.java

```diff
@@ -71,82 +69,29 @@ public abstract class PhantomCleanable<T> extends PhantomReference<T>
      * @param referent the referent to track
      * @param cleaner  the {@code Cleaner} to register with
      */
+    @SuppressWarnings("this-escape")
     public PhantomCleanable(T referent, Cleaner cleaner) {
         super(Objects.requireNonNull(referent), CleanerImpl.getCleanerImpl(cleaner).queue);
-        this.list = CleanerImpl.getCleanerImpl(cleaner).phantomCleanableList;
-        insert();
+        index = -1;
+        list = CleanerImpl.getCleanerImpl(cleaner).activeList;
+        list.insert(this);
+
+        // Check that list insertion populated the backlinks.
+        assert node != null;
+        assert index >= 0;
 
         // Ensure referent and cleaner remain accessible
         Reference.reachabilityFence(referent);
         Reference.reachabilityFence(cleaner);
     }
 
-    /**
-     * Construct a new root of the list; not inserted.
-     */
-    PhantomCleanable() {
-        super(null, null);
-        this.list = this;
-    }
-
-    /**
-     * Insert this PhantomCleanable after the list head.
-     */
-    /*[IF CRIU_SUPPORT]*/
-    @NotCheckpointSafe
-    /*[ENDIF] CRIU_SUPPORT */
-    private void insert() {
-        synchronized (list) {
-            prev = list;
-            next = list.next;
-            next.prev = this;
-            list.next = this;
-        }
-    }
-
-    /**
-     * Remove this PhantomCleanable from the list.
-     *
-     * @return true if Cleanable was removed or false if not because
-     * it had already been removed before
-     */
-    /*[IF CRIU_SUPPORT]*/
-    @NotCheckpointSafe
-    /*[ENDIF] CRIU_SUPPORT */
-    private boolean remove() {
-        synchronized (list) {
-            if (next != this) {
-                next.prev = prev;
-                prev.next = next;
-                prev = this;
-                next = this;
-                return true;
-            }
-            return false;
-        }
-    }
-
-    /**
-     * Returns true if the list's next reference refers to itself.
-     *
-     * @return true if the list is empty
-     */
-    /*[IF CRIU_SUPPORT]*/
-    @NotCheckpointSafe
-    /*[ENDIF] CRIU_SUPPORT */
-    boolean isListEmpty() {
-        synchronized (list) {
-            return list == list.next;
-        }
-    }
-
     /**
      * Unregister this PhantomCleanable and invoke {@link #performCleanup()},
      * ensuring at-most-once semantics.
      */
     @Override
     public final void clean() {
-        if (remove()) {
+        if (list.remove(this)) {
```

Comment:
Does this method not need `@NotCheckpointSafe` (for CRIU)? `remove()` had the annotation before, now the synchronization is in `list.remove()` - same comment for `clear()`. Or should the annotation be attached to `CleanerImpl.CleanableList.remove()`?

Reply:
The `@NotCheckpointSafe` annotation was added according to the method in a stacktrace from `JVMCheckpointException` or `JVMRestoreException` when CRIU checkpoint or restore failed at single-thread mode. Usually the failure is intermittent. 
The original remove() was removed, it doesn't mean `CleanerImpl.CleanableList.remove()` needs the annotation as well.
The annotation will be added if a `JVMCRIUException` has the method in question. Yeah, it is added as per need.

Reply:
I guess we'll see if `cmdLineTester_criu_nonPortableRestore` start failing again.